<HTML><HEAD>
<TITLE>MGI Editorial Interface Bug Tracking - November, 1998</TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF">

<CENTER>
<H1>MGI Editoral Interface Bug Tracking - November, 1998</H1>
<H4>Click RELOAD on your browser to make sure you are looking at the latest version.</H4>
</CENTER> <HR NOSHADE>

<OL>
<LI>Refer to the <A HREF="MEI.html">MGD Requirements</A> for specific information 
    on the behavior of the MGI interface.
<LI>Send mail to <A HREF="mailto:lec@informatics.jax.org">Lori</A>
	to report a bug in the MGI interface.  Please do not send bug
	reports to <CODE>info-se</CODE>
</OL>
<HR NOSHADE>

<TABLE>
<TR>
<TD bgcolor="#00FF00">&nbsp;&nbsp;<TD>= Fixed
<TD bgcolor="#FFFF00">&nbsp;&nbsp;<TD>= In progress
<TD bgcolor="#FF0000">&nbsp;&nbsp;<TD>= Not started
<TD bgcolor="#CC9900">&nbsp;&nbsp;<TD>= On hold
<TD bgcolor="#CC9999">&nbsp;&nbsp;<TD>= No bug
<TD bgcolor="#FFFFFF">&nbsp;&nbsp;<TD>= Modification Request
</TABLE>

<p>

<H4>Pre-Production - November 2-November 12, 1998</H4>
<TABLE BORDER=1>
<TR ALIGN=left><TH>Bug<TH>Date<TH>Status<TH>Contact<TH>Notes

<TR bgcolor="#00FF00">
	<TD>Processing of multiple new symbols for non-mouse species causes error.
	    <CODE>@maxMarker2</CODE> not declared.
	<TD>11/02/1998; plg/sfr
	<TD>Fixed
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#00FF00">
	<TD>After a conflict was reported in Homology when entering the <I>Tk</I> marker,
	    all of the records for the <I>Tk</I> homology were deleted and had to be readded.
	<TD>11/03/1998; plg
	<TD>Fixed
	<TD>Lori
	<TD>The conflict is reported and deleted if the record is being ADDED.  But, if the record
	    is being MODIFIED, then no deletes occur and the user must fix the conflict manually.

<TR bgcolor="#00FF00">
	<TD>When adding a new tissue record, after the add has completed the interface does not show
	    the new tissue as the currently selected record in the Search Results.
	<TD>11/03/1998; let
	<TD>Fixed
	<TD>Lori
	<TD>There is inconsistent behavior in MGD and GXD in regards to Adds.  The Antigen, Antibody, Image, Index/Stage 
	    forms in GXD have been re-implemented to remain consistent with MGD.

<TR bgcolor="#00FF00">
	<TD>The data in the development database is recent as of October 30.  The snapshot was to be as
	    of October 24.
	<TD>11/03/1998; let
	<TD>Fixed
	<TD>Lori
	<TD>The development database was accidently overwritten.  A backup will be restored so
	    that testing can resume as planned tomorrow morning.

<TR bgcolor="#CC9999">
	<TD>In the Tissue screen, searches for Data Sets associated with a particular tissue seem to
	    be taking a very long time to process.
	<TD>11/03/1998; let
	<TD>No bug
	<TD>Lori
	<TD>This is a known performance problem due to the number of Molecular Segments in the database.

<TR bgcolor="#00FF00">
	<TD>After entering the J: value in the Reference screen on the Molecular Probes
	    form, the citation does not appear.
	<TD>11/03/1998; neb
	<TD>Fixed
	<TD>Lori
	<TD>The incorrect widget was being accessed during verification.

<TR bgcolor="#00FF00">
	<TD>No permissions on GXD Antigen table.
	<TD>11/03/1998; let
	<TD>Fixed
	<TD>Lori
	<TD>John to re-run permission scripts.

<TR bgcolor="#00FF00">
	<TD>
	In processing Current Contents #42 (1st pass) (Step #2 of the Test Plan
	for CC Update); the following message appeared:
	<CODE>
	[nlm.py -Ujeo -P/home/jeo/.mgi_password -c /home/jeo/CC#42 ]
	<BR>
	Traceback (innermost last):
	<BR>
	File "nlm.py", line 83, in ?
	<BR>
          import mgdlib
	  <BR>
          ImportError: No module named mgdlib
	  <BR>
          PROCESSING COMPLETED
	  <BR>
	</CODE>
	<TD>11/04/1998; jeo
	<TD>Fixed
	<TD>Lori
	<TD>Missing <CODE>/home/lec/lib/mgdpython</CODE> in PYTHONPATH in run script.

<TR bgcolor="#CC9999">
	<TD>Added a new Marker <I>Tgl4</I> with J: 503.  Looks like the citation field
	    is being populated from the title field?
	<TD>11/04/1998; djr
	<TD>No bug
	<TD>Lori
	<TD>The Authors field for J:503 contains "Y linked zinc finger protein 1Y linked zinc finger protein 1Mark WH;
	    Signorelli K; Blum M; Kwee L; Lacy E".  This needs to be fixed.

<TR bgcolor="#CC9999">
	<TD>Is the ability to enter new Marker Alleles via GXD part of this release?
	<TD>11/04/1998; let
	<TD>No bug
	<TD>Lori
	<TD>Yes.  The documentation needs to be updated and permissions granted for GXD editors to insert records into
	    the Marker Allele table.

<TR bgcolor="#CC9999">
	<TD>In the History table on the Marker form, the 'Name' and 'Citation' fields look punctuated
	    with a period  (.) when something is entered in them.... see new marker Krn3.
	<TD>11/04/1998; djr
	<TD>No bug
	<TD>Lori
	<TD>The period is the Table widget's mechanism for signalling the user that the total length of the
	    string exceeds what is visible on the screen at one time.

<TR bgcolor="00FF00">
	<TD>Current Contents Update
	<CODE>
	PROCESSING...
	<BR>
	[nlm.py -Ujeo -P/home/jeo/.mgi_password -c /home/jeo/CC#42 ]
	<BR>
	Error: Could not open file nlm.journals.seen
	<BR>
	PROCESSING COMPLETED
	</CODE>
	<TD>11/04/1998; jeo
	<TD>Fixed
	<TD>Lori
	<TD>Missing <CODE>nlm.jouranls.seen</CODE> and <CODE>cc.journals</CODE> from test directory.

<TR bgcolor="#00FF00">
	<TD>The GXD assay screen is not correctly verifying whether or not a probe is
	    already cross-referenced to a particular marker.
	    <CODE>
	    Marker = Hgl
	    <BR>
	    Probe = MGI:1298423 or MGI:23707
	    <BR>
	    Error:  "The Marker 'Hgl' is not cross- referenced to this Probe Accession ID."
	    </CODE>
	<TD>11/04/1998; let
	<TD>Fixed
	<TD>Lori
	<TD>The verification routine was using the wrong table to perform the verification.

<TR bgcolor="#00FF00">
	<TD>Multiple errors when processing the broadcast.
	<TD>11/04/1998; djr
	<TD>Fixed
	<TD>Lori
	<TD>Errors in <CODE>broadcast.py</CODE>.

<TR bgcolor="#CC9999">
	<TD>User cannot print out diagnostic reports for Current Contents update.
	    The file is highlighted but the message says file: /home/jeo was printed.
	    <CODE>/usr/ucb/lpr:diva: unknown printer/class</CODE>
	<TD>11/04/1998; jeo
	<TD>No bug
	<TD>Lori
	<TD><CODE>argus</CODE> does not know of the printer <CODE>diva</CODE>.  Choose printer
	    <CODE>bentley</CODE> instead during testing.

<TR bgcolor="#00FF00">
	<TD>During add of primary markers in Mapping form, the message:
	<CODE>Required Field:  "ExptTypeMenu"</CODE> appears and the user is not
	transferred to the secondary form.
	<TD>11/04/1998; plg
	<TD>Fixed
	<TD>Lori
	<TD>The test directory did not have the most current versions of the PCD files.

<TR bgcolor="#CC9999">
	<TD>Tried transfering data belonging to Amph J28029 to Bis2. Should have
	transfered Homology, MLC, Mapping and Probes Amph1: cDNA1. From editing
	screen looks like the  process happens but data is not transfered; nor
	does mail appear to have been generated to affected editors.
	<TD>11/04/1998; djr
	<TD>No bug?
	<TD>Lori
	<TD>Could not determine problem.  Is now working.  May have been an old version in test directory?

<TR bgcolor="#CC9999">
	<TD>Homology testing:  When I enter new homolgy and the assay type and then
	add as in J10996 the assay was conserved location. after it added the
	data, unreviewed also was selected as well as conserved location.
	<TD>11/04/1998; plg
	<TD>No bug
	<TD>Lori
	<TD>Did the user place X's in the appropriate cells to assign the Assay?
	    Cannot find a problem.  No further reports of this problem.

<TR bgcolor="#00FF00">
	<TD>The system allows me to enter an Age Range when entering no Age value and
	    adding a new Specimen row.  Shouldn't this be disallowed?
	<TD>11/04/1998; dab
	<TD>Fixed
	<TD>Lori
	<TD>Yes.  Age ranges are not valid for the default of "Not Specified".

<TR bgcolor="#00FF00">
	<TD>The requirements document says that the species default = <I>Not Specified</I> in
	the Antigen and Antibody forms.  However, the form is defaulting to <I>mouse, laboratory</I>.
	<TD>11/04/1998; let
	<TD>Fixed
	<TD>Lori
	<TD>Documentation has been fixed.  Default is <I>mouse, laboratory</I>.

<TR bgcolor="#00FF00">
	<TD>During add of Marker Allele via GXD interface, the Allele Name is defaulting to
	    <I>targeted mutation</I>.  Debbie will have to be notified later to enter the
	    appropriate name.  What should we do?
	<TD>11/04/1998; let
	<TD>Fixed
	<TD>Lori
	<TD>Will allow GXD Editor to add the Allele name at the time the Allele Symbol is added.
	    Will use a default Name of <I>allele name in progress</I>.  Added segment to existing QCR 
	    "Markers Which Contain Alleles" to report these <I>in progress</I> Allele Names.

<TR bgcolor="#CC9999">
	<TD>Transfered data from J10794 of Cas1 to Bub1.. had mapping and
	    probes. Odd thing in probes... it had compiled data from the transfered
	    Jnum 10794 and another ref J17819. It looks like the Marker associated with
	    J17819 also changed to Bub1...doesn't seem like that should happen should
	    it..should J17819 still be related to Cas1?
	<TD>11/04/1998; djr
	<TD>No bug
	<TD>Lori
	<TD>During a Marker Transfer, if a Marker/J: combination exists in a Probe record, then the
	    entire Probe is transferred to the new Marker even if the Probe contains other References which
	    are not designated during the transfer.

<TR bgcolor="#00FF00">
	<TD>I am unable to select an experiment type after I have entered and added
	    on the primary screen. I get a report which says can not modify
	    experiment type.  Example J15048.
	<TD>11/04/1998; plg
	<TD>Fixed
	<TD>Lori
	<TD>Program was detecting a modification to the Experiment Type when there was none.

<TR bgcolor="#00FF00">
	<TD>After making changes to an existing MLC entry and choosing Clear, the page
	does not clear and the edits made to the record remain.  There is no way to get
	out of the record other than quitting the program.  Upon re-entering, the original
	record is back.
	<TD>11/05/1998; rmb
	<TD>Fixed
	<TD>Lori
	<TD>Missing appropriate callback for OK button; left out during conversion.

<TR bgcolor="#00FF00">
	<TD>When trying to add a Probe, the message:
	<BR><CODE> "A molecular segment is Currently Selected"</CODE> 
	appears when trying to include the parent clone information in the Parent Clone Accession ID field.
	<TD>11/05/1998; let
	<TD>Fixed
	<TD>Lori
	<TD>Duplicate object names in TeleUSE hierarchies causing problems.  Had to remove duplicate ID labels.

<TR bgcolor="#00FF00">
	<TD>The date modified for the image editing screen is not being updated when a
	record is modified. See J:49133.
	<TD>11/05/1998; let
	<TD>Fixed
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#00FF00">
	<TD>RI screen: When designation is selected from the RI set, the Animal field is not populated.
	<TD>11/05/1998; plg
	<TD>Fixed
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#CC9999">
	<TD>I just tried to copy a note in an in situ results field (using F3), I then
	clicked OK on the note screen and attempted to click Save Results on the
	results screen. I got this status report: "Implicit converstion from
	datatype 'TEXT' to 'VARCHAR' is not allowed. Use the CONVERT function to run
	this query."
	<TD>11/05/1998; let
	<TD>No bug
	<TD>Lori
	<TD>Notes cannot exceed 255.

<TR bgcolor="#CC9999">
	<TD>Down and Up arrow keys do not advance the MLC Reference box.
	<TD>11/05/1998; rmb
	<TD>No bug
	<TD>Lori
	<TD>See the Requirements document section on Table Processing for new behavior.

<TR bgcolor="#00FF00">
	<TD>The created & modified dates in the Anatomical Dictionary screen do not
	    match the format found in the rest of the editing screens.
	<TD>11/05/1998; let
	<TD>Fixed
	<TD>Lori
	<TD>Geoff fixed this.

<TR bgcolor="#FFFFFF">
	<TD>Prevent MLC transfer from executing if no current MLC record.
	<TD>11/05/1998; lec
	<TD>Modification Request
	<TD>Geoff
	<TD>&nbsp;

<TR bgcolor="#00FF00">
	<TD>GXD Assay form not displaying side panel correctly.
	<TD>11/05/1998; let
	<TD>Fixed
	<TD>Lori
	<TD>Broke during last installation...how?  CVGel panel was not unmanaged upon entry.

<TR bgcolor="#CC9999">
	<TD>When trying to use an Antibody where "recognizes product in Western Blot" is 
	    Not Specified, the Assay screen will not allow the user to use the Antibody.
	<TD>11/05/1998; dab
	<TD>No bug
	<TD>Lori
	<TD>The interface alerts the user, but doesn't seem to disallow the use of the Antibody.
	    Need further feedback from Dale.

<TR bgcolor="#CC9999">
	<TD>After attaching an Antibody to an Assay record and modifying the attributes of the
	    Antibody (changing the Marker symbol, etc.), the user is still able to use the
	    Antibody even though it is now contradictory to the attributes of the Assay.
	<TD>11/05/1998; dab
	<TD>No bug
	<TD>Lori
	<TD>Some data verifications simply alert the user and do not prevent the use of the
	    contradictory data.  Subsequent changes to Antibodies, etc. are not re-verified
	    against existing Assays.  QCRs can be used to report such contradictions.

<TR bgcolor="#CC9999">
	<TD>User added 2 structures to the AD clipboard, but 3 structures appear in the AD list
	    on the Assay form.
	<TD>11/05/1998; dab
	<TD>No bug?
	<TD>Lori
	<TD>User may have double-clicked on tree structure in AD form?

<TR bgcolor="#00FF00">
	<TD>Getting "incorrect syntax" message when display InSitu Results dialog.
	<TD>11/06/1998; dab
	<TD>Fixed
	<TD>Lori
	<TD>Wrong object name used.

<TR bgcolor="#00FF00">
	<TD>Search not working correctly for wildcard Marker values in Marker, Homology, MLDP, Molecular forms.
	<TD>11/06/1998; lec
	<TD>Fixed
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#FFFFFF">
	<TD>When selecting a Primary Marker for the Experiment Marker list, can the Experiment Marker
	    list scroll so that the selected Marker is visible?
	<TD>11/06/1998; plg
	<TD>Modification Request
	<TD>Lori
	<TD>User request.

<TR bgcolor="#FFFFFF">
	<TD>Send MLC Transfer email message to Richard Baldarelli
	<TD>11/06/1998; lec, rmb
	<TD>Modification Request
	<TD>Lori
	<TD>User request.

<TR bgcolor="#FFFF00">
	<TD>Broadcast - notify Richard Baldarelli of MLC status.
	<TD>11/06/1998; rmb
	<TD>In progress
	<TD>Lori
	<TD>User request.

<TR bgcolor="#FFFFFF">
	<TD>QCR report to check MLC Image index file against current Nomenclature
	<TD>11/06/1998; rmb
	<TD>Modification Request
	<TD>Lori
	<TD>User request.

<TR bgcolor="#00FF00">
	<TD>"donkey" and "sheep" Species being displayed as "pig, domestic" and "provirus" on Web.
	<TD>11/06/1998; dab
	<TD>Fixed
	<TD>Lori
	<TD>Error in default values within GEI was adding incorrect values into DB table field.
	    Fixed in Production as well.

<TR bgcolor="#CC9999">
	<TD>During Marker Allele Merge, after entering the Old and New Allele symbols, the message:
	<BR>
	<CODE>Old Allele Symbol required during this merge</CODE>
	<TD>11/06/1998; djr
	<TD>No bug?
	<TD>Lori
	<TD>User did not TAB out of Old Allele field properly?

<TR bgcolor="#00FF00">
	<TD>During Marker Transfer, check GXD Assay/Antibody records. Send Marker Transfer report
	    to Laura Trepanier.
	<TD>11/06/1998; djr
	<TD>Fixed
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#00FF00">
	<TD>User cannot modify MLC Class data.
	<TD>11/06/1998; rmb
	<TD>Fixed
	<TD>Lori
	<TD>Richard Baldarelli needs permissions to edit the <I>MRK_Classes</I> table.

<TR bgcolor="#FFFFFF">
	<TD>When changing Marker Chromosome from known to another known, can the Offset values be re-set to -1.0?
	    Also, verify that Marker is not an Anchor Locus.
	<TD>11/09/1998; djr/lec
	<TD>Modification Request
	<TD>Lori
	<TD>User request.

<TR bgcolor="#FFFFFF">
	<TD>Remove default of Marker Type to <I>Gene</I> in Marker form.
	<TD>11/09/1998; djr
	<TD>Modification Request
	<TD>Lori
	<TD>User request.

<TR bgcolor="#FFFFFF">
	<TD>Remove check of Offset values when entering Marker Aliases.
	<TD>11/09/1998; djr
	<TD>Modification Request
	<TD>Lori
	<TD>User request.

<TR bgcolor="#FFFFFF">
	<TD>Add MLC Transfer Status report to QCR page.
	<TD>11/09/1998; lec
	<TD>Modification Request
	<TD>Lori
	<TD>Report was in Nightly Report script but not on QCR page.

<TR bgcolor="#FFFFFF">
	<TD>MGI: is listed before J: in Accession table on References form.
	    Can the J: be listed first?
	<TD>11/10/1998; mgm/jeo
	<TD>Modification Request
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#FFFFFF">
	<TD>QCR to check GXD Assay Probe/Markers which no longer are valid.
	<TD>11/10/1998; let
	<TD>Modification Request
	<TD>Lori
	<TD>Installed on QCR page.

<TR bgcolor="#FFFFFF">
	<TD>Always force Status Report message box to front of screen.
	    It gets hidden and the interface freezes up.
	<TD>11/10/1998; djr
	<TD>Modification Request
	<TD>Lori
	<TD>When any form is brought into focus, the Status Message will be brought to the front.

<TR bgcolor="#00FF00">
	<TD>If I delete a record in mapping example J49740, everything deletes
	    except for the primary markers list to the side of the main entry
	    screen. I have to clear to delete them.
	<TD>11/10/1998; plg
	<TD>Fixed
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#00FF00">
	<TD>Added an Article using term <I>Abstr</I> in the Page field but the
	    Review Status did not default to <I>Unreviewed</I>.
	<TD>11/10/1998; jeo
	<TD>Fixed
	<TD>Lori
	<TD>Review Status default is now case-insensitive.

<TR bgcolor="#CC9999">
	<TD>Tabbing in Strain form freezes screen.
	<TD>11/10/1998; mlp
	<TD>No bug?
	<TD>Lori
	<TD>Cannot re-create bug; is working fine.

<TR bgcolor="#00FF00">
	<TD>During processing of CC#43:
	<BR>
	<CODE>
	File "nlm.py", line 401, in processSO
	<BR>
	[part1, part2] = string.split(rec['SO'], ';', 1);
	<BR>
	ValueError: unpack list of wrong size
	<BR>
	PROCESSING COMPLETED
	</CODE>
	<TD>11/12/1998; jeo/mgm
	<TD>Fixed
	<TD>Lori
	<TD>File has <CODE>SO  - Arch Virol 1998:259-279</CODE> format (no volume or issue).
	    Program was not handling this format.

<TR bgcolor="#FFFFFF">
	<TD>Report duplications during Strain Merge, Strain Add, Strain Update.
	<TD>11/10/1998; lec/mlp
	<TD>Modification Request
	<TD>Lori
	<TD>Strain Merge duplicates will abort transaction.
	    <BR>
	    Strain Add/Update duplicates will display a warning message.

</TABLE>

<H4>Production - Installed November 13, 1998</H4>
<TABLE BORDER=1>
<TR ALIGN=left><TH>Bug<TH>Date<TH>Status<TH>Contact<TH>Notes

<TR bgcolor="#00FF00">
	<TD>Text not wrapping in Experiment Text field.
	<TD>11/13/1998; plg
	<TD>Fixed
	<TD>Lori
	<TD>The <CODE>wordWrap</CODE> and <CODE>scrollHorizontal</CODE> attributes for
	    <CODE>Notes</CODE> were not set properly.

<TR bgcolor="#00FF00">
	<TD>Error when adding a Sequence Accession number during a Probe Reference add.
	<TD>11/13/1998; neb
	<TD>Fixed
	<TD>Lori
	<TD>The <CODE>ACCRef_process</CODE> stored procedure had a <CODE>select @accKey</CODE>
	    line in it which was causing the new Probe Reference key to be overwritten in the
	    Editing Interface.

<TR bgcolor="#00FF00">
	<TD>Clearing of screen/lists in Molecular Segments not working properly upon delete.
	<TD>11/13/1998; lec
	<TD>Fixed
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#00FF00">
	<TD>Display of Probe Reference record after entering RFLP information incorrect.
	<TD>11/13/1998; neb
	<TD>Fixed
	<TD>Lori
	<TD>New Probe Reference key was being overwritten by new Allele key and new RFLP key.

<TR bgcolor="#FFFFFF">
	<TD>Automatically add the Probe Reference record during add of new GXD J#/Probe combination.
	<TD>11/13/1998; let/lec
	<TD>Modification request
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#FFFFFF">
	<TD>Make <CODE>jaguar</CODE> the default printer instead of <CODE>zinger</CODE>.
	<TD>11/16/1998; plg
	<TD>Modification request
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#FFFFFF">
	<TD>Change <CODE>$HOME/mgdreport</CODE> to <CODE>$HOME/mgireport</CODE>
	    and automatically create the directory (if it doesn't exist) when the user 
	    logs into the interface.
	<TD>11/16/1998; lec
	<TD>Modification request
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#00FF00">
	<TD>When adding J:39566, RI Data, a status report says
	   <CODE>column alleleLine in table MLD_RIData does not allow null values</CODE>
	<TD>11/16/1998; lec
	<TD>Fixed
	<TD>Lori
	<TD>Haplotype rows tagged for Adds when Primary Markers are selected.  The allele
	    line value needs to be checked as well as the row's edit mode to prevent additions 
	    where the allele line is not entered.

<TR bgcolor="#00FF00">
	<TD>Error "Must declare variable @key" when trying to modify Probe <I>Pk3 cDNA2</I>
	    marker from <I>Pk3 </I>to <I>Pdha1.</I>
	<TD>11/16/1998; neb
	<TD>Fixed
	<TD>Lori
	<TD>Note SQL command not being re-initialized.

<TR bgcolor="#00FF00">
	<TD>When Accession numbers are moved to a new Marker during a Nomenclature event,
	    the non-MGI Accession numbers are being moved with <CODE>preferred = 0</CODE>.
	<TD>11/16/1998; lec
	<TD>Fixed
	<TD>Lori
	<TD>The procedure <CODE>MRK_copyAcc</CODE> needs to check the logical DB and set
	    the preferred bit to 1 for non-MGI Acccession numbers.
	    <BR>
	    The procedure <CODE>MRK_updateKeys</CODE> should only set preferred = 0
	    for MGI Accession numbers.  All others, leave preferred as is.

<TR bgcolor="#00FF00">
	<TD>Interface crashing after tabbing through Reference Year field.
	<TD>11/16/1998; mgm
	<TD>Fixed
	<TD>Lori
	<TD>The <CODE>mgi_year</CODE> routine was passing a <CODE>const char*</CODE>
	    which was then being modified within the routine.
	    <BR><BR>
	    Still crashing.  Remove call to translation <CODE>VerifyReferenceDate</CODE>
	    to see if problem goes away.
	    <BR><BR>
	    Removing callback <CODE>VerifyReferenceDate</CODE> seems to have fixed
	    this problem.  More investigation needed as to why this callback is
	    causing a core dump.

<TR bgcolor="#00FF00">
	<TD>Searching for J: in Molecular Segment form takes too long.
	<TD>11/16/1998; neb
	<TD>Fixed
	<TD>Lori
	<TD>Improve search by using identifier key instead of J: field.

<TR bgcolor="#CC9999">
	<TD>After entering Haplotype data in a Cross (J:49736), the message
	    <BR>
	    <CODE>Number of Haplotypes does not match number of Markers</CODE>
	    is displayed.  There are 7 markers and 7 haplotypes.
	<TD>11/17/1998; plg
	<TD>No bug
	<TD>Lori
	<TD>The <CODE>Matrix</CODE> columns for all 7 markers are not equal to <I>yes</I>,
	    so the verification is working correctly.

<TR bgcolor="#00FF00">
	<TD>When tabbing through a blank Assay Type in MLDP, many error messages appear.
	<TD>11/17/1998; lec
	<TD>Fixed
	<TD>Lori
	<TD>Verification needs to check for value of length 0 or value = " ".

<TR bgcolor="#00FF00">
	<TD>Override of J: during Reference add is being ignored.
	<TD>11/17/1998; mgm
	<TD>Fixed
	<TD>Lori
	<TD>When J: and MGI: rows were reversed, the appropriate change was not propagated
	    to the Add processing in the Reference module.

<TR bgcolor="#00FF00">
	<TD>Broadcast lock file is not being removed.
	<TD>11/18/1998; djr
	<TD>Fixed
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#00FF00">
	<TD>Submission Accession numbers are not displaying in Reference form.
	<TD>11/18/1998; jjm
	<TD>Fixed
	<TD>Lori
	<TD>View used to retrieve Accession numbers is ignoring Submission references.

<TR bgcolor="#00FF00">
	<TD><CODE>Matrix?</CODE> is defaulting to <I>no</I>; change to <I>yes</I>.
	<TD>11/18/1998; plg
	<TD>Fixed
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#FFFFFF">
	<TD>When entering a new journal <I>NMR Biomed</I>, the message
	<BR><CODE>No Alternative Found</CODE>
	<BR>appeared and wiped out the Journal field.  Can this behavior be changed?
	<TD>11/18/1998; jeo
	<TD>Modification Request
	<TD>Lori
	<TD>&nbsp;

<TR bgcolor="#CC9999">
	<TD>When generating a Probe report for J:47964, the message says
	    <BR><CODE>_Refs_key = J:48292</CODE>, but the correct results appear
	    in the report.
	<TD>11/19/1998; neb
	<TD>No bug
	<TD>Lori
	<TD><CODE>_Refs_key = 48292</CODE> is the unique identifier for <CODE>J:47964</CODE>.

<TR bgcolor="#00FF00">
	<TD>Attempt to insert duplicate records into PRB_RFLV table.
	<TD>11/23/1998; neb
	<TD>Fixed
	<TD>Lori
	<TD>If same Endonuclease/Marker used, then same RFLV record key should be used.

<TR bgcolor="#00FF00">
	<TD>When adding a Probe record and deleting an erroneous Marker, syntax error occurs.
	<TD>11/23/1998; neb
	<TD>Fixed
	<TD>Lori
	<TD>User should not use <I>Delete Row</I> if Marker has not previously been added.
	    Modify code to check for valid Marker before constructing edit statements for Marker
	    rows.

<TR bgcolor="#00FF00">
	<TD>When adding a new Strain during entry of RFLP data, the message:
	    <CODE>Must declare variable '@key'</CODE> appears.
	<TD>11/30/1998; neb
	<TD>Fixed
	<TD>Lori
	<TD><CODE>@key</CODE> variable not being declared in <I>VerifyStrains</I>.

<TR bgcolor="#00FF00">
	<TD>When modify Mapping Haplotype data, the message:
	    <CODE>Invalid column 'haplotypes'</CODE> appears.
	<TD>11/30/1998; plg
	<TD>Fixed
	<TD>Lori
	<TD>Should be <CODE>alleleLine</CODE>.

<TR bgcolor="#00FF00">
	<TD>When modify Mapping Haplotype data, the message:
	    <CODE>Invalid column 'mice'</CODE> appears.
	<TD>11/30/1998; plg
	<TD>Fixed
	<TD>Lori
	<TD>Should be <CODE>offspringNmbr</CODE>.

<TR bgcolor="#CC9999">
	<TD>When entering a Genotype, received the message:
	    <CODE>incorrect syntax</CODE>
	<TD>11/30/1998; dab
	<TD>No bug?
	<TD>Lori
	<TD>No <CODE>_Strain_key</CODE> during Genotype insert.  Did user forget to
	    tab out of Strain field?  Should put in a fix to use Not Specified if
	    Strain key is empty...

<TR bgcolor="#00FF00">
	<TD>When trying to remove a Parent Clone Accession number from a record,
	    <CODE>incorrect syntax</CODE>
	<TD>12/03/1998; djr
	<TD>Fixed
	<TD>Lori
	<TD>The code needs to check if the new Object ID is blank or NULL.

<TR bgcolor="#00FF00">
	<TD>When duplicating a GXD Gel record, the original record is returned and
	    loaded into the Search Results list.
	<TD>12/03/1998; let
	<TD>Fixed
	<TD>Lori
	<TD>The Duplicate function needs to re-initialize the primary key before
	    performing the duplication so that the new record key is placed into
	    the top->ID widget.

<TR bgcolor="#00FF00">
	<TD>When selecting a non-Anonymous library in the Molecular Probe form,
	    the select Library information is not displayed correctly.
	<TD>12/15/1998; lec
	<TD>Fixed
	<TD>Lori
	<TD>The Library List was refering to <CODE>ID</CODE> when it should refer to
	    <CODE>SourceID</CODE>.

<TR bgcolor="#CC9999">
	<TD>When modifying the parental number in the 2X2 table for J:49653, Cross 3, the message:
	    <CODE>Duplicate Order in Table.  Cannot Modify.</CODE> appears.  The modification
	    was successful, however.
	<TD>12/15/1998; plg
	<TD>No bug
	<TD>Lori
	<TD>Error in ordering of Statistics records.  This error was cleared up when the
	    Statistics were re-generated during the record modification.

</TABLE>

<H4>Comments</H4>
<OL>
<LI>Many bugs which were found during Editor testing should have been caught during SE testing.
    <BR>Conclusions:  more thorough SE testing needed.
<P>
<LI>Bugs which were found after the Production release should have been caught
    during testing.  
    <BR>Conclusions:  more thorough testing needed, particularly in Molecular Probes & Segments and Mapping.
<P>
<LI>New test cases will be added to the appropriate Test Plans in response to the
    detected bugs.
<P>
<LI>Development database was in the incorrect state when testing started and 
    permissions were missing.  Environment variables missing.
    <BR>
    Conclusions:  We need to pay more attention to these details before formal testing begins.
<P>
<LI>Users very helpful in suggesting alternative behaviors, enhancements, etc.
    All but one enhancement was implemented during testing, but this may not always be
    possible or feasible.
<P>
<LI>Core files are generated if the user performs a search which brings back more search results
    than the interface can handle.  The application eventually crashes due to lack of memory.
    We could put in a limit of the number of rows returned.
</OL>

</BODY></HTML>
