<HTML>
<HEADER>
<TITLE>
MGI HOMOLOGY LOADS
</TITLE>
</HEADER>

<BODY>
 
<H2>MGI HOMOLOGY LOADS</H2>

<H4>Introduction</H4>
MGD stores Marker information and Accession numbers for non-mouse Markers which are
used in the Homology portion of the database.
<P>
From the Web Homology detail page, MGD links to species-specific databases (SSDBs) either 
by Marker symbol or by SSDB Accession number.  Not every SSDB provides Accession numbers
for their Marker objects.
<P>
SSDBs are also used to maintain MGD nomenclature and mapping information for non-mouse markers.
<P>
The Marker information and Accession numbers for the non-mouse Markers must be updated
on a regular basis to reflect the SSDB changes in nomenclature, mapping position, etc.

<H4>Getting the Data Dumps</H4>

The species-specific Accession numbers are loaded into MGD by an automated process 
which uses data dumps provided by the SSDBs.
<P>

<TABLE BORDER=1>
<TR ALIGN=left><TH>SSDB<TH>Data File<TH>Contact<TH>Frequency of Data Load
<TR><TD>LocusLink
    <TD><A HREF="http://prodwww/~lec/mgireport/locuslinkload/output/LL.out">LL.out</A>
	<BR>Data file is automatically extracted from the FTP site <CODE>ftp.ncbi.nlm.nih.gov</CODE>
	    to <CODE>titan:/usr/local/mgi/locuslinkload/output/</CODE>.
    <TD>Donna Maglott <BR>(maglott@ncbi.nlm.nih.gov)
    <TD>Daily

<TR><TD>RATMAP
    <TD><A HREF="http://prodwww/~lec/mgireport/ratmapload/RatURL.txt">RatURL.txt</A>
	<BR>Data file is automatically extracted from the FTP site <CODE>genmac17.gen.gu.se</CODE>
	    to <CODE>titan:/usr/local/mgi/ratmapload/output/</CODE>.
    <TD>Fredrik Stahl <BR>(Fredrik.Stahl@gen.gu.se)
    <TD>Daily (TR 555; RATMAP file is only updated once a month)

<TR><TD>GDB
    <TD><A HREF="http://prodwww/~lec/mgireport/HOMOLOGY/Gene.medlineid.tab">Gene.medlineid.tab</A>
	<BR>Data file is automatically extracted from the FTP site <CODE>syb.hsc.gdb.org</CODE>
	    to <CODE>titan:/usr/local/mgi/ei/mgd/mgireport/HOMOLOGY</CODE>.
    <TD>Connie Talbot <BR>(cct@gdb.org)
    <TD>Daily (GDB file is only updated once a month)

</TABLE>

<H4>LocusLink</H4>

We are currently using LocusLink to update MGD human symbols, mapping and accession information.
We are only loading the GDB ID from the LocusLink file.

<OL>
<LI>All MGD Human symbols which do not contain a GDB ID are updated with the correct
    GDB ID based on a symbol match between MGD and the LocusLink symbol (official or interim).
<P>
<LI>All MGD Human symbols are updated from the LocusLink file based on a GDB ID match
    between MGD and LocusLink. The fields which are updated are:
	<UL>
	<LI>symbol
	<LI>name
	<LI>chromosome
	<LI>offset
	</UL>
<P>
<LI>We output the following:
	<UL>
	<LI>MGD Human Symbols updated today (LL.found.sql.rpt)
	<LI>MGD Human Symbols w/out a match in LocusLink (LL.notfound.sql.rpt)
	</UL>
<P>
<LI>Executed from a <CODE>sybase</CODE> cron on <CODE>titan</CODE>:
	<PRE>
	/usr/local/mgi/locuslink/LLupdate.sh -l
	</PRE>
</OL>

<H4>RATMAP</H4>

We are currently using RATMAP to update MGD rat mapping and accession information.
We are not updating MGD Rat symbols from RATMAP.

<OL>
<LI>All MGD Rat symbols which do not contain a RATMAP ID are updated with the correct
    RATMAP ID based on a symbol match between MGD and the RATMAP symbol.
<P>
<LI>All MGD Rat symbols are updated from the RATMAP file based on a RATMAP ID match
    between MGD and RATMAP. The fields which are updated are:
	<UL>
	<LI>chromosome
	<LI>offset
	</UL>
<P>
<LI>We output the following:
	<UL>
	<LI>MGD Rat Symbols updated today (RATMAP.found.sql.rpt)
	<LI>MGD Rat Symbols w/out a match in RATMAP (RATMAP.notfound.sql.rpt)
	<LI>Conflicts between MGD Rat Symbols and RATMAP Symbols (RATMAP.conflicts.sql.rpt)
	<LI>Potential Mouse/Rat Matches (RATMAP.potentials.sql.rpt)
	<LI>RATMAP Symbols and GenBankIDs (RATMAP.genbankIDs.sql.rpt)
	</UL>
<P>
<LI>In addition, we retrieve the following files from the RATMAP ftp site:
	<UL>
	<LI>RATMAP Homolog file (RATMAP.Homolog.txt)
	</UL>
<P>
<LI>Executed from a <CODE>sybase</CODE> cron on <CODE>titan</CODE>:
	<PRE>
	/usr/local/mgi/ratmapload/RATMAPupdate.sh -l
	</PRE>
</OL>

<H4>GDB</H4>

We are currently using GDB to create a report of potential human homologies based
on matching GDB citations with MGD citations by symbol for which no human/mouse
ortholog exists in MGD.

<OL>
<P>
<LI>We output the following:
	<UL>
	<LI>MGD Entries with Potential Human Homologies (HMD_Potentials.rpt)
	</UL>
<P>
<LI>Executed from a <CODE>sybase</CODE> cron on <CODE>titan</CODE>:
	<PRE>
	/usr/local/mgi/qcreports_db/GDBreport.sh -f
	</PRE>
</OL>

<HR>

</BODY>
</HTML>
