<HTML>
<HEADER>
<TITLE>
MGD EDITING INTERFACE: Design and Requirements - Molecular Probes & Segments
</TITLE>
</HEADER>

<BODY>

<H2>
MGD EDITING INTERFACE: Design and Requirements - Molecular Probes & Segments
</H2>
<BR>

<H4>Introduction</H4>
Molecular Segments contains:
	<UL>
	<LI>cDNA
	<LI>genomic
	<LI>DNA (construct)
	<LI>mitochondrial
	<LI>oligo
	<LI>primer
	</UL>
<P>
All Molecular Segments which are not designated as <I>primers</I> have a <I>Library</I> record which defines:
	<UL>
	<LI>Species
	<LI>Strain
	<LI>Tissue
	<LI>Cell Line
	<LI>Age
	<LI>Sex
	<LI>Tissue Treatment
	</UL>
<P>
Molecular Segments have <I>relationships</I> with Markers:
	<UL>
	<LI><B>A</B>, a Marker <B>amplifies</B> a primer
	<LI><B>E</B>, a Marker <B>encodes</B> a probe if it hybridizes 100% to the probe
	<LI><B>H</B>, a Marker <B>hybridizes</B> to a probe
	<LI><B>M</B>, designation for <B>MIT</B> primers
	<LI><B>P</B>, a <B>putative</B> association derived from dbEST GenBank IDs
	</UL>
<P>
Each Molecular Segment may have a list of References which may contain RFLP information, including
nucleotide (GenBank) accession numbers.
<P>
Molecular Segments are loaded from external sources as well as manual input.  These updates occur
on a regular basis (quarterly, monthly, etc.):
	<UL>
	<LI>MIT primers
	<LI>I.M.A.G.E. clones
	</UL>

<H4>Editing Form</H4>
<OL>
<LI><A HREF="molsegprobe.gif">Probe GIF image</A>
<LI><A HREF="molsegprimer.gif">Primer GIF image</A>
<LI><A HREF="molsegreference.gif">Molecular Segment Reference GIF image</A>
<LI><A HREF="molseglibrary.gif">Molecular Source Library GIF image</A>
</OL>

<H4>Accession Numbers</H4>
<OL>
<LI>MGI:, system-generated
<LI>WashU, cross reference to WashU
<LI>I.M.A.G.E., cross reference to I.M.A.G.E. consortium
<LI>Sequence, cross reference to GenBank, EMBL, DDBJ (nucleotide sequences)
<LI>ATCC, cross reference to American Type Culture Collection
</OL>

<H4>Data Verifications</H4>
<OL>
<LI>Sequence Type (see <A HREF="MEI.html#dataVerOption">Data Verifications section on Option Menus</A>)
	<UL>
	<LI>cDNA
	<LI>genomic
	<LI>DNA (construct)
	<LI>mitochondrial
	<LI>oligo
	<LI>primer
	</UL>
<P>
<LI>Vector Type (see <A HREF="MEI.html#dataVerOption">Data Verifications section on Option Menus</A>)
	<UL>
	<LI>YAC
	<LI>Cosmid
	<LI>Plasmid
	<LI>Phage
	<LI>BAC
	<LI>P1
	<LI>Phagemid
	</UL>
<P>
<LI>J Numbers (see <A HREF="MEI.html#dataVerJnumber">Data Verifications section on J Numbers</A>)
<P>
<LI>Markers (see <A HREF="MEI.html#dataVerMarker">Data Verifications section on Markers</A>)
<P>
<LI>Parent Clone Accession ID (see <A HREF="MEI.html#dataVerAcc">Data Verifications section on Accession Numbers</A>)
<P>
<LI>Library (see <A HREF="MEI.html#dataVerLookup">Data Verifications section on Lookup Lists</A>)
    <P>dbEST EST libraries are downloaded during the dbEST EST load
<P>
<LI>Species (see <A HREF="MEI.html#dataVerOption">Data Verifications section on Option Menus</A>)
<P>
<LI>Strain (see <A HREF="MEI.html#dataVerStrain">Data Verifications section on Strain</A>)
<P>
<LI>Tissue (see <A HREF="MEI.html#dataVerTissue">Data Verifications section on Tissue</A>)
<P>
<LI>Age (see <A HREF="MEI.html#dataVerAge">Data Verifications section on Age</A>)
<P>
<LI>Sex (see <A HREF="MEI.html#dataVerOption">Data Verifications section on Option Menus</A>)
	<UL>
	<LI>Female
	<LI>Male
	<LI>Pooled
	</UL>
<P>
</OL>

<H4>Fields</H4>
As the user selects Sequence Type from the option menu, the middle of the form will change depending on the
Sequence Type selected and different fields will become available for viewing/editing.
<UL>
<LI>Required
	<OL>
	<LI>Sequence Type
	<LI>Name
	<LI>Non-Primers Only
		<OL>
		<LI>Vector Type
		<LI>Library
		<LI>Species
		<LI>Strain
		<LI>Tissue
		<LI>Age
		<LI>Sex
		</OL>
	</OL>
<P>
<LI>Not Required
	<OL>
	<LI>Region Covered
	<LI>Non-Primers Only
		<OL>
		<LI>Parent Clone Accession ID
		<LI>Parent Clone
		<LI>Insert Site
		<LI>Insert Size
		<LI>Cell Line
		<LI>Tissue Treatment
		</OL>
	<LI>Primers Only
		<OL>
		<LI>Sequence 1
		<LI>Sequence 2
		<LI>Repeat Unit
		<LI>More Than 1 Product
		<LI>Product Size
		</OL>
	<LI>Marker
	<LI>Notes
	</OL>
</UL>

<H4>Defaults</H4>
<OL>
<LI>Species; default is <I>mouse, laboratory</I>
<LI>Vector Type; default is <I>Not Specified</I>
<LI>Age; default is <I>Not Specified</I>
<LI>Sex; default is <I>Not Specified</I>
<LI>If <I>Tissue</I> is <I>unfertilized egg</I>, then <I>Age</I> default is <I>Not Applicable</I>
<LI>If <I>Age</I> is <I>Not Applicable</I>, then <I>Tissue</I> must be <I>unfertilized egg</I>
</OL>

<H4>Deletions</H4>
There are no restrictions on deleting a Molecular Probe & Segment.

<H4>Library</H4>
A Library selection populates the Library text field and performs a lookup of:
	<UL>
	<LI>Species
	<LI>Strain
	<LI>Tissue
	<LI>Cell Line
	<LI>Age
	<LI>Sex
	<LI>Tissue Treatment
	</UL>
<P>
To edit a non-anonymous Library, the user can also use the Molecular Probe Libraries form.
To access this form, choose 
<A HREF="molseglibrary.gif"><I>Edit->Molecular Probe Libraries</I></A>.
<P>

<H4>Miscellaneous</H4>
<OL>
<LI>To view/edit Molecular Segment master records, set the <I>References</I> toggle OFF.
<P>
<LI>To view/edit Molecular Segment References records, set the <I>References</I> toggle ON.
<P>
<LI>The Marker/Probe relationship (<I>R</I> column in the Marker table):
	<UL>
	<LI>must be <I>H</I> (hybridizes) for Probes of non-mouse source.
	    This includes <I>Not Specified</I> and <I>Not Applicable</I> sources.
	<LI>can only be <I>P</I> during an EST bulk load.
	<LI>can be updated to another relationship from <I>P</I>
	</UL>
<P>
<LI>The <I>KFMemorial</I> button places the following text in the <I>Notes</I> field:
	<PRE>This clone is derived from a gene that produces alternate transcripts.</PRE>
<P>
<LI>Strains in the RFLV table on the Reference form must be entered in the following format:
	<PRE> Strain1, Strain2, Strain3</PRE>
    <P>The delimiter "comma space" <B>MUST</B> be used to separate two distinct Strains.
       If the correct delimiter is not used, then the Strains will be treated as one Strain.
<P>
<LI>Tabbing through the <I>Endo</I> and <I>Marker</I> columns in the RFLV table will cause the
    previous row's values to be copied to the current row for those columns.
<P>
<LI>The deletion of a master Molecular Segment record will delete the Molecular Segment and all
    of its References.
<P>
<LI>Editing of WashU clone records added from the electronic dbEST load is restricted.  These 
    electronically loaded records are distinguished by the reference J:57656.  Users may
    add new Accession numbers to any WashU clone record, but they may not edit Accession numbers,
    clone name or source information for any WashU clone which is referenced by J:57656.
    <P>
    <UL>
    <LI>If the user attempts to edit the Accession numbers of a WashU clone record from the dbEST load, 
        the following error message will appear:
	<PRE>
	Accession ID is referenced by J:57656, WashU-HHMI Database Download.
	Cannot edit.
	</PRE>
    <LI>If the user attempts to edit the Source information of a WashU clone record from the dbEST load, 
        the following error message will appear:
	<PRE>
	Molecular Segment is referenced by J:57656, WashU-HHMI Database Download.
	Cannot modify the Source information.
	</PRE>
    <LI>If the user attempts to edit the name of a WashU clone record from the dbEST load, 
        the following error message will appear:
	<PRE>
	Molecular Segment is referenced by J:57656, WashU-HHMI Database Download.
	Cannot modify the name.
	</PRE>
    </UL>
<P>
</OL>

<H4>Triggers/Views/Stored Procedures for Molecular Segment tables (<I>PRB_</I>)</H4>
<OL>
<LI>Triggers ($SYBASE/admin/triggers):  <I>PRB.tr</I>, <I>PRB.drop</I>
<LI>Views ($SYBASE/admin/views):  <I>PRB.view</I>
<LI>Stored Procedures ($SYBASE/admin/procedures):  <I>PRB.pr</I>
</OL>

<H3>RELATED DOCUMENTS</H3>
<OL>
<LI><A HREF = "MEICV.html">Lookups (controlled vocabulary)</A>
</OL>

</BODY>
</HTML>
