<HTML>
<HEADER>
<TITLE>
MGI EDITING INTERFACE: Design and Requirements - Orthology
</TITLE>
</HEADER>

<BODY BGCOLOR="#FFFFFF">

<H2>
MGI EDITING INTERFACE: Design and Requirements - Orthology
</H2>
<BR>

<H4>Introduction</H4>
Orthology defines relationships of genes among different mammalian organism.  This interface allows the
user to define a Orthology reported in a specific Reference.  Once that Orthology has been entered, the
interface will attempt to merge that Orthology with other Orthologies which have at least one of the
symbol/organism pairs in common.  This creates a Orthology "class" thru transitivity:

	<PRE>
	If Mouse A is orthologous to Human A and Human A is orthologus to Rat A, 
	then Mouse A is orthologous to Rat A
	</PRE>

<P>
Orthologies are defined by the Assay(s) which were used to determine the orthology.  Therefore, one
Reference may report 3 different orthologous relationships using 3 different assays.  The 3 orthologies
may fall under the same orthology "class", or they may not.
<P>
In this interface, it is important to remember that one selected record is equivalent to one defined orthology and
not an entire orthology class.
<P>
Downloads from Organism-specific databases are used to update the Symbol attributes for non-mouse organism on
a regular basis.
<P>

<H4>Editing Form</H4>
<A HREF="orthology.gif">GIF image</A>

<H4>Accession Numbers</H4>
<OL>
<LI>MGI:; system-generated
<LI>LocusLink accession numbers have no prefix; cross referenced to LocusLink
<LI>RGD accession numbers have no prefix; cross referenced to RGD
</OL>

<H4>Data Verifications</H4>
<OL>
<LI>Organism (see <A HREF="MEI.html#dataVerPulldown">Data Verifications section on Pulldown Menus</A>)
<P>
<LI>Chromosome (see <A HREF="MEI.html#dataVerChromosome">Data Verifications section on Non-Mouse Marker Chromosomes</A>)
<P>
<LI>Assays (see <A HREF="MEI.html#dataVerLookup">Data Verifications section on Controlled Vocabulary Lists</A>)
<P>
</OL>

<H4>Fields</H4>
<UL>
<LI>Required
	<OL>
	<LI>J:
	<LI>Organism
	<LI>Symbol
	<LI>Chr
	<LI>Name
	<LI>Assay
	</OL>
<P>
<LI>Not Required
	<OL>
	<LI>Cyto
	<LI>Acc ID
	<LI>Notes
	</OL>
</UL>

<H4>Defaults</H4>
The Marker table <B>always</B> displays the following organism:
<OL>
<LI>mouse, laboratory
<LI>human
<LI>rat
</OL>

When a orthology record is displayed, all of the default organism plus all of the organism
defined in the orthology are displayed.  If no Symbol is associated with a default organism,
then that organism is not part of the orthology record.
<P>
Default organism values cannot be overwritten by any other organism value.  The row of a
default organism value is reserved solely for that organism.

<H4>Searches</H4>
All fields in the editing form are searchable with the following exceptions:
<OL>
<LI>Numbered columns in the Marker table
<LI>Numbered columns in the Assay table
</OL>

<H4>Deletions</H4>
There are no restrictions on deleting a Orthology.

<H4>Modifying Symbols</H4>
Symbols for mouse can only be modified using the Marker form.
<P>
To edit symbols for non-mouse organism, use the Marker NonMouse editing form which is 
available from the <I>Edit->Marker Information</I> selection on the Orthology editing form.  
See the <A HREF="MEIMarkerNonMouse.html"> MGI EDITING INTERFACE: Design and Requirements - Non-Mouse Marker</A> document for details about this editing form.
<P>

<H4>Adding/Modifying a Orthology</H4>
<OL>
<LI>Enter all of the Organism/Symbols in the Symbol table.
<P>
<LI>Select all of the Assays which are used in defining the orthologies for the Symbols.
<P>
<LI>Enter all of the Notes which are available.
<P>
<LI>Select the Symbol/Assay combinations which define each orthology by clicking in the
    Assay column which corresponds to the <I>No.</I> column of the Symbol.
<P>
<LI>A Organism/Symbol can belong to more than one Assay group, but must belong to at least one
    Assay group or else it will be dropped from the orthology.  Therefore, to delete a Symbol 
    from a orthology, de-select the Symbol from the Assay to which it is attached.
    <P>
    If a new Assay is to be assigned to a pre-existing Orthology, insert the new Assay
    adjacent to the pre-existing Orthology.  This will ensure that the Assay groups are
    assigned to the same Orthology key.
<P>
<LI>After any add/modification, the system will attempt to "merge" the current orthology record
    with other orthologies which belong to the same "class", based on the Symbols/Organism defined.
<P>
<LI>As Organism/Symbol are entered in the Symbol table, the system will detect any collisions,
    i.e. the entry of a Organism/Symbol which conflicts with a previously defined orthology.
    <P>
    For example, if a orthology class exists for <I>mouse, Ren1</I> and <I>human, REN</I> and
    the user attempts to enter <I>mouse, Ren1</I> and <I>human, REN1</I> and the symbol
    <I>human, REN1</I> exists in the database, then the system will detect a conflict and 
    report this to the user.
    <P>
<P>
<LI>If a Organism/Symbol needs to be replaced with a differenct symbol for a given Orthology Class,
    then the entire Orthology Class must be deleted and re-added.
<P>
</OL>

<H4>Notes</H4>
Orthology notes are attached to unique orthology records.  A unique orthology record is defined
by the Organism to which the Assay is assigned.
<P>
At the present time, the user cannot attach notes to a specific orthology reference or to a
specific orthology marker symbol.

<H4>Miscellaneous</H4>
<OL>
<LI>Setting the <I>Display Inferred Orthologies</I> ON before a <I>Search</I> will retrieve all
    Orthology records belonging to the same Orthology Class of the Orthology record which matches
    the user's search criteria.
    <P>
    For example, searching for <I>mouse, laboratory/Ren1</I> with the toggle OFF will retrieve
    8 Orthology records where <I>mouse, laboratory/Ren1</I> is a member of each Orthology record.
    <P>
    Searching for <I>mouse, laboratory/Ren1</I> with the toggle ON will retrieve all of the
    16 Orthology records which belong to the Orthology Class of which <I>mouse, laboratory/Ren1</I>
    is a member.  <I>mouse, laboratory/Ren1</I> will not necessarily be a member of every
    Orthology record which is retrieved, however.
<P>
<LI>The <I>Class ID</I> for every Orthology record is displayed for the user.  The user can also
    search for a specific <I>Class ID</I>.  However, <I>Class IDs</I> are assigned by the system
    and cannot be modified by the user.
<P>
<LI>Certain Organism's Symbols are updated from data downloads of organism-specific databases.
<P>
</OL>

<H3>RELATED DOCUMENTS</H3>
<OL>
<LI><A HREF = "MEICV.html">Controlled Vocabulary</A>
</OL>

</BODY>
</HTML>
